File: SwiftMarshal/RubyString.swift
//
//  RubyString.swift
//  SwiftMarshal
//
//  Created by Jonas Zaugg on 25.02.15.
//  Copyright (c) 2015 Jonas Zaugg. All rights reserved.
//

import Foundation

public class RubyString: RubyType {
    
    final private var _str:String
    final var str:String {
        get {
            return _str
        }
    }
    
    public init(str: String) {
        self._str = str
    }
    
    public override func inspect(inout sb: NSMutableString, indent: Int) {
        sb.append("\"")
        sb.append(str)
        sb.append("\"")
    }
    
    /*public func hashCode() -> Int {
    var hash = 3
    hash = 53 * hash + Arrays.hashCode(self.bytes)
    }*/
    
    public override func isEqual(object: AnyObject?) -> Bool {
        if let object:AnyObject = object {
            if object.isKindOfClass(RubyString) {
                return (object as RubyString) == self
            } else {
                return false
            }
        } else {
            return false
        }
    }
    
    /*@Override
    public boolean equals(Object obj) {
    if (obj == null) {
    return false;
    }
    if (getClass() != obj.getClass()) {
    return false;
    }
    final RubyString other = (RubyString) obj;
    if (!Arrays.equals(this.bytes, other.bytes)) {
    return false;
    }
    return true;
    }*/
    
}

func ==(lhs: RubyString, rhs: RubyString) -> Bool {
    return lhs.str == rhs.str
}

File: SwiftMarshal/Marshal Ruby-Java/types/RubyString.java
/*
 * Copyright 2012 mfizz.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.mfizz.ruby.types;

/*
 * #%L
 * mfz-ruby-marshal
 * %%
 * Copyright (C) 2012 mfizz
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */

import java.util.Arrays;

/**
 * Strings in Ruby are just an array of byte's.  Unfortunately very dangerous
 * to just directly convert into Java strings since the encoding is unknown.
 * 
 * Immutable instance of a Ruby string.
 * 
 * @author mfizz
 */
public class RubyString extends RubyType {
    
    final private byte[] bytes;

    public RubyString(byte[] bytes) {
        this.bytes = bytes;
    }

    public byte[] getBytes() {
        return bytes;
    }
    
    @Override
    public void inspect(StringBuilder sb, int indent) {
        try {        
            sb.append("\"");
            sb.append(new String(this.bytes, "UTF-8"));
            sb.append("\"");
        } catch (Exception e) {
            throw new RuntimeException("utf-8 charset doesn't exist?");
        }
    }

    @Override
    public int hashCode() {
        int hash = 3;
        hash = 53 * hash + Arrays.hashCode(this.bytes);
        return hash;
    }

    @Override
    public boolean equals(Object obj) {
        if (obj == null) {
            return false;
        }
        if (getClass() != obj.getClass()) {
            return false;
        }
        final RubyString other = (RubyString) obj;
        if (!Arrays.equals(this.bytes, other.bytes)) {
            return false;
        }
        return true;
    }
    
}
